### ISSUE TITLE
plotting with index of objects should use `FixedLocator`

### ISSUE THREAD

**Issue by tacaswell** [pandas-dev/pandas#7612]
As pointed out here (https://stackoverflow.com/questions/24468333/set-xlim-for-pandas-matplotlib-where-index-is-string/24477138#24477138) if the index of a data frame is objects (that is not, the index obviously convertible to scalars and the data is being plotted against `range(len(df))`), the ticklabels are set to be string representations of the objects (I presume via `set_xticklabels`) but the locator is still a `AutoLocator` which means if you change the xlimits or pan/zoom the axes, the ticklabels become de-coupled from the data.

A solution is to also use a fixed locator which will pin the ticks to only be on the index locations.

```
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker

df = pd.DataFrame({'Foo': pd.Series([2,3,4], index=['2002', '2003', '2004'])})
fig, ax = plt.subplots()

df.plot(ax=ax)

ax.xaxis.set_major_locator(mticker.FixedLocator(np.arange(len(df))))
ax.xaxis.set_major_formatter(mticker.FixedFormatter(df.index))
```

I can make changes to the code if someone points me to where the relevant section is.

**Comment by jreback** [comment]
closing as stale. pls reopen if still an issue.

**Comment by tacaswell** [comment]
This is still a bug and should be re-opened.

**Comment by jorisvandenbossche** [comment]
When someone tries to fix this, make sure to test whether #2980, #11465 and #15912 are then solved.

@TomAugspurger @sinhrks do you have an idea where this should best be handled? Somewhere in `_make_plot`?

**Comment by TomAugspurger** [comment]
Yeah, `_make_plot` is where the change is needed, but if we're fixing it in `MPLPlot`, then we'll need to do some refactoring.

**Comment by firasm** [comment]
FYI: This still appears to be an issue and drove me crazy till I found this!

**Comment by nrebena** [comment]
So I think i found the place where to put the change suggested by @tacaswell, and it fix #7612, #15912, #22334.
Regarding  issues #2980 and #11465, I do not think they are related to this issue.
